src/frontend/hooks/useLocation.ts

```diff
@@ -0,0 +1,113 @@
+// so we need to check it.
+const LOCATION_TIMEOUT = 10000;
+
+export function useLocation(options: LocationOptions) {
+  const {minTimeInterval: timeInterval = 200, ...debounceOptions} = options;
+  const [location, setLocation] = React.useState<LocationObject | undefined>();
+
+  const [permissions] = useForegroundPermissions();
+
+  useFocusEffect(
+    React.useCallback(() => {
+      if (!permissions || !permissions.granted) return;
+
+      let ignore = false;
+      const LocationSubscriptionProm = watchPositionAsync(
+        {accuracy: Accuracy.BestForNavigation},
+        debounceLocation({
+          ...debounceOptions,
+        })(location => {
+          if (ignore) return;
+          setLocation(location);
+        }),
+      );
+
+      return () => {
+        ignore = true;
+        LocationSubscriptionProm.then(sub => sub.remove());
+      };
+    }, [permissions]),
+  );
+
+  return {location, latLon: location ? getCoords(location) : undefined};
+}
+
+function debounceLocation({
+  maxDistanceInterval,
+  minDistanceInterval = 1,
+  maxTimeInterval = 1000,
+}: Omit<LocationOptions, 'minTimeInterval'>) {
+  let lastLocation: LocationObject | undefined;
+  let ruler: any;
+  let rulerLat = 0;
+  let timeout: ReturnType<typeof setTimeout>;
+  return function (callback: (location: LocationObject | undefined) => any) {
+    return function (location: LocationObject) {
+      // The user can turn off location services via the quick settings dropdown
+      // (swiping down from the top of their phone screen) without moving away
+      // from the app. In this case the location will just stop updating and we
+      // won't know why. If we haven't had a location update for a while, we check
+      // on the provider status to see if location services are enabled, so that
+      // we can update the state with the current status
+      if (timeout) clearTimeout(timeout);
+      timeout = setTimeout(async () => {
+        if (!(await hasServicesEnabledAsync())) {
+          lastLocation = undefined;
+          callback(undefined);
+          return;
+        }
+      }, LOCATION_TIMEOUT);
+      if (!lastLocation) {
+        lastLocation = location;
+        callback(location);
+        return;
+      }
+      const coords = getCoords(location);
+      const lastCoords = getCoords(lastLocation);
+      const rulerLatDisplacement = Math.abs(rulerLat - coords[1]);
+      if (!ruler || rulerLatDisplacement > MAX_RULER_LAT_DISPLACEMENT) {
+        rulerLat = coords[1];
+        ruler = new CheapRuler(rulerLat, 'meters');
+      }
+      const distance = ruler.distance(coords, lastCoords);
+      const time = location.timestamp - lastLocation.timestamp;
+      if (
+        distance > maxDistanceInterval ||
+        (time > maxTimeInterval && distance > minDistanceInterval)
+      ) {
+        lastLocation = location;
+        callback(location);
+      }
+    };
+  };
+}
+
+function getCoords(location: LocationObject) {
```

Comment:
more useful to work with a tuple in this case (TS infers `number[]` without this)

```suggestion
function getCoords(location: LocationObject): [number, number] {
```

Reply:
https://github.com/digidem/CoMapeo-mobile/pull/107/commits/fa942647c4ca4cb1be9dfbf9cc3a53aed71d531a